### PR TITLE
Feat/Pagination

### PR DESCRIPTION
**Pull request by gromkii**
Closes [PATH-585](https://gohypergiant.atlassian.net/browse/PATH-585)

## ✅ Pull Request Checklist

- [ ] Included link to corresponding [GitHub Issue](https://github.com/gohypergiant/standard-toolkit/issues).
- [x] The commit message follows [conventional commit](https://www.conventionalcommits.org/en/v1.0.0/#summary) [extended](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type) guidelines.
- [x] Added/updated unit tests and storybook for this change (for bug fixes / features).
- [x] Added/updated documentation (for bug fixes / features)
- [x] Filled out test instructions.
- [x] Added changeset (for bug fixes / features).

## 📝 Test Instructions

- `pnpm --filter=@accelint/design-toolkit run preview`
- Navigate to `Pagination` component, observe functionality.

## ❓ Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: design-toolkit/components/src/components/pagination/index.tsx [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+  pageCount,
+  onChange,
+  ...rest
+}: BasePaginationProps) {
+  return (
+    <Pagination.Provider value={{ currentPage, pageCount }}>
+      <div
+        className={container({ className: classNames?.container })}
+        {...rest}
+      >
```
Comment (colehart) [inline design-toolkit/components/src/components/pagination/index.tsx]: consider adding `role='navigation'` and an `aria-label` to the container here for better accessibility.

File: design-toolkit/components/src/components/pagination/pagination.test.tsx [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+function setup({ currentPage, pageCount }: Partial<BasePaginationProps> = {}) {
+  render(<Pagination currentPage={currentPage} pageCount={pageCount} />);
+}
+
+describe('Pagination', () => {
+  it('should render', () => {
+    setup({ currentPage: 1, pageCount: 5 });
+    expect(screen.getByText('1')).toBeInTheDocument();
+  });
```
Comment (colehart) [inline design-toolkit/components/src/components/pagination/pagination.test.tsx]: not sure how robust the other component tests are for the toolkit since we have storybook integration testing, and if we want to beef this up, consider adding tests for:
- page navigation
- edge cases (first/last page, single page)
- disabled state logic
- range calculation function

File: design-toolkit/components/src/components/pagination/index.tsx [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+  // Below max display.
+  if (pageCount < 5) {
+    return {
+      minRange: 1,
+      maxRange: pageCount,
+    };
+  }
+
+  // Middle.
+  if (currentPage >= 3 && currentPage <= pageCount - 3) {
```
Comment (colehart) [inline design-toolkit/components/src/components/pagination/index.tsx]: not sure, but should this be instead:

```suggestion
  if (currentPage >= 3 && currentPage < pageCount - 2) {
```

double check current logic when we have large page counts (like 10 for example), and you're getting near the end of the list. if it's wrong for [5, 6, 7, 8, 9], try the suggested logic instead

Comment (gromkii) [reply design-toolkit/components/src/components/pagination/index.tsx]: Updated the logic here.

File: design-toolkit/components/src/components/pagination/index.tsx (hunk 1 above)
Comment (gromkii) [reply design-toolkit/components/src/components/pagination/index.tsx]: Added an aria-label here, but the linter has suggested using the semantic `nav` tag over `div` over using role, so will be implementing that.

File: design-toolkit/components/src/components/pagination/index.tsx [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+        <Pagination.Previous
+          onPress={() => onChange?.(currentPage - 1)}
+          className={classNames?.controls}
+        />
+        <Pagination.NumberContainer
+          onPress={(nextPage) => onChange?.(nextPage)}
+          className={classNames?.pages}
+        />
+        <Pagination.Next
+          onPress={() => onChange?.(currentPage + 1)}
```
Comment (colehart) [inline design-toolkit/components/src/components/pagination/index.tsx]: should validate the boundary before mathing

```suggestion
onPress={() => {
  if (currentPage > 1) onChange?.(currentPage + 1);
}}
```

File: design-toolkit/components/src/components/pagination/types.ts [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+
+import type { PropsWithChildren } from 'react';
+
+export type BasePaginationProps = PropsWithChildren & {
+  pageCount?: number;
+  currentPage?: number;
```
[the remark this answers is not in the document]
Comment (colehart) [reply design-toolkit/components/src/components/pagination/types.ts]: if we wanted to add an isLoading prop / state for when the `pageCount` / `currentPage` data isn't available yet (like while fetching), that might be helpful for users.

**Review by colehart — COMMENTED** [review]
general ideas that might be out of scope / may warrant a future ticket:

- Add page size selector (common UX pattern)
```
<Pagination.PageSize 
  options={[10, 25, 50, 100]} 
  value={pageSize}
  onChange={setPageSize}
/>
```

- Add "Go to page" input for direct navigation
`<Pagination.JumpToPage onSubmit={handleJump} />`

**Comment by gromkii** [comment]
> general ideas that might be out of scope / may warrant a future ticket:
> 
> * Add page size selector (common UX pattern)
> 
> ```
> <Pagination.PageSize 
>   options={[10, 25, 50, 100]} 
>   value={pageSize}
>   onChange={setPageSize}
> />
> ```
> 
> * Add "Go to page" input for direct navigation
>   `<Pagination.JumpToPage onSubmit={handleJump} />`

I agree that these would be helpful features for a future iteration! As per discussion with design, we're wanting to limit the scope of this current version to display 5 pages at a time.

File: design-toolkit/components/src/components/pagination/types.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (gromkii) [reply design-toolkit/components/src/components/pagination/types.ts]: These should be required, you are correct. I dig the idea of a loading state for this component, and for the moment I'll include it as a prop. I might be helpful to find a solution for a visual representation for that in a future iteration, but should be helpful enough.

File: design-toolkit/components/src/components/pagination/index.tsx [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+
+/**
+ *
+ * Pagination - A lightweight implementation for page navigation.
+ *
+ * @example
+ * <Pagination currentPage={1} pageCount={5} onChange={handleOnChange} />
+ */
+export function Pagination({
+  children,
```
[the remark this answers is not in the document]
Comment (gromkii) [reply design-toolkit/components/src/components/pagination/index.tsx]: Removing children for now as it's not being used in this first iteration.

File: design-toolkit/components/src/components/pagination/index.tsx [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+    !(pageCount && currentPage) ||
+    currentPage > pageCount ||
+    pageCount < 1 ||
+    currentPage < 1
+  ) {
+    return { minRange: 0, maxRange: 0 };
+  }
+
+  // Below max display.
+  if (pageCount < 5) {
```
Comment (switzerb) [inline design-toolkit/components/src/components/pagination/index.tsx]:
```suggestion
  if (pageCount < DEFAULT_MAX_RANGE) {
```

I see in the ticket requirements we are only doing the 5 page display, so I get why this is here and I don't consider it a blocker -- but every engineering instinct I have tells me not to hard-code these numbers.

File: design-toolkit/components/src/components/pagination/types.ts [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+
+import type { PropsWithChildren } from 'react';
+
+export type BasePaginationProps = PropsWithChildren & {
```
Comment (colehart) [inline design-toolkit/components/src/components/pagination/types.ts]: bc we removed the children prop recently, this type can be updated to not include the `PropsWithChildren` helper type here.

**Review by kalisjoshua — DISMISSED** [review]
